In [3]:
import requests
import json
import pandas as pd
import tweepy

In [4]:
with open(r'C:\Users\roast\Desktop\twitter_keys.txt', 'r') as file:
    tokens = file.readlines()
API_KEY = tokens[0]
API_SECRET = tokens[1]
ACCESS_KEY = tokens[2]
ACCESS_SECRET = tokens[3]
bearer = tokens[4]

In [5]:
auth = tweepy.OAuthHandler(API_KEY, API_SECRET)
auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)
client = tweepy.Client(bearer_token = bearer)

In [140]:
def df_response(res):
    tw_users1 = {tw_user['id']: tw_user for tw_user in res.includes['users']}
    df = pd.DataFrame()

    for tweet in res.data:
        if tw_users1[tweet.author_id]:
            tw_user = tw_users1[tweet.author_id]
        temp_dict = {
            'tweet_id': [tweet.id],
            'username': [tw_user.username],
            'text': [tweet.text],
            'created_at': [tweet.created_at],
            'location': [tw_user.location],
            'source': [tweet.source],
            'url': ["https://twitter.com/"+str(tw_user.username)+"/status/"+str(tweet.id)]
        }
        temp_df = pd.DataFrame(temp_dict)
        df = pd.concat([temp_df, df])

    return df

In [145]:
#Creating loop for word: mitaka
mitaka_df = pd.DataFrame()
query1 = '(#mitaka OR mitaka) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = 1555697979629006848 #8/5/22 7:55pm EST

for i in range(4):
    tweets1 = client.search_recent_tweets(query=query1, max_results = 100, until_id = last_tw_id, 
                                        user_fields = ['username', 'location'],
                                        tweet_fields=['created_at', 'source'],
                                        expansions = ['author_id'])
    temp_df = df_response(tweets1).sort_values(['created_at'], ascending=False)
    last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
    
    mitaka_df = pd.concat([temp_df, mitaka_df])

In [147]:
mitaka_df.shape

(364, 7)

In [151]:
mitaka_df.head()

tweet_id       username  \
0  1553777739773186048      Vulpeqxla   
0  1553776451186458625      aokisotaa   
0  1553775430276194304  retromanjapan   
0  1553773879650394115      nymphahri   
0  1553756835047690240       moonrcin   

                                                text  \
0  i need to draw mitaka again soon or else i wil...   
0  finally decided to check out the denji edit ta...   
0  The Showa Retro In https://t.co/H3BK93ITy6 #wo...   
0  Asa Mitaka ⛓️\nShould I make the actual cospla...   
0  mitaka best girl \n\n#チェンソーマン #chainsawman htt...   

                 created_at      location               source  \
0 2022-07-31 16:20:55+00:00   they/them ♥   Twitter for iPhone   
0 2022-07-31 16:15:48+00:00  19, they/she  Twitter for Android   
0 2022-07-31 16:11:44+00:00  Tokyo, Japan    Revive Social App   
0 2022-07-31 16:05:35+00:00        Italia  Twitter for Android   
0 2022-07-31 14:57:51+00:00            🇵🇭   Twitter for iPhone   

                                                 url  
0  https://twitter.com/Vulpeqxla/status/155377773...  
0  https://twitter.com/aokisotaa/status/155377645...  
0  https://twitter.com/retromanjapan/status/15537...  
0  https://twitter.com/nymphahri/status/155377387...  
0  https://twitter.com/moonrcin/status/1553756835...

In [161]:
#Creating loop for word: asa mitaka
asamitaka_df = pd.DataFrame()
query2 = '(asamitaka OR "asa mitaka" OR #asamitaka) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = 1555671453357510659

for i in range(2):
    tweets2 = client.search_recent_tweets(query=query2, max_results = 100, until_id = last_tw_id,
                                    user_fields = ['username', 'location'],
                                    tweet_fields=['created_at', 'source'],
                                    expansions = ['author_id'])
    temp_df = df_response(tweets2).sort_values(['created_at'], ascending=False)
    last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
        
    asamitaka_df = pd.concat([temp_df, asamitaka_df])

In [162]:
asamitaka_df.shape

(155, 7)

In [163]:
asamitaka_df.head()

tweet_id        username  \
0  1554121032348442627        olya_fry   
0  1554113023526445057    currynasimba   
0  1554099904741785601    penumbrateaa   
0  1554096146527780864  anothergiselle   
0  1554082224487952384          jhksy_   

                                                text  \
0  Look at me\n\n#chainsawman #チェンソーマン #asamitaka...   
0  NOBODY TALK SHIT ABOUT ASA MITAKA, THIS IS AN ...   
0  I believe in Asa Mitaka supremacy https://t.co...   
0  Asa Mitaka sketches I did the other day https:...   
0                                  i want asa mitaka   

                 created_at                location               source  \
0 2022-08-01 15:05:02+00:00  Somewhere in the Earth     Twitter for iPad   
0 2022-08-01 14:33:13+00:00            Indonesia 🇲🇨   Twitter for iPhone   
0 2022-08-01 13:41:05+00:00                ur heart  Twitter for Android   
0 2022-08-01 13:26:09+00:00      vzla (arepa frita)      Twitter Web App   
0 2022-08-01 12:30:50+00:00                   lia 🤍   Twitter for iPhone   

                                                 url  
0  https://twitter.com/olya_fry/status/1554121032...  
0  https://twitter.com/currynasimba/status/155411...  
0  https://twitter.com/penumbrateaa/status/155409...  
0  https://twitter.com/anothergiselle/status/1554...  
0  https://twitter.com/jhksy_/status/155408222448...

In [170]:
#Creating loop for word: mitaka asa
mitakaasa_df = pd.DataFrame()
query3 = '(mitakaasa OR "mitaka asa" OR #mitakaasa) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = 1555670538278047744

for i in range(1):
    tweets3 = client.search_recent_tweets(query=query3, max_results = 100, until_id = last_tw_id,
                                    user_fields = ['username', 'location'],
                                    tweet_fields=['created_at', 'source'],
                                    expansions = ['author_id'])
    temp_df = df_response(tweets3).sort_values(['created_at'], ascending=False)
    last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
        
    mitakaasa_df = pd.concat([temp_df, mitakaasa_df])

In [173]:
mitakaasa_df.shape

(14, 7)

In [175]:
mitakaasa_df.head()

tweet_id        username  \
0  1555552823143710721        gojotoji   
0  1555515203839983616  Kshatriya_666_   
0  1555430172043939841        Skullse1   
0  1555241643611406336        Skullse1   
0  1555213967684841474    Jackburger__   

                                                text  \
0   mitaka asa seems so cool, I need to read csm pt2   
0  Mitaka using Yusuke Murata's style\n.\n.\n#cha...   
0  "wanna be my boyfriend?"\n\n#mitaka_asa\n#chai...   
0  WIP\n\n"What weapon will I be?"\n#mitaka_asa \...   
0  🔥🔥war devil \n\n#doodle #art #digitaldrawing #...   

                 created_at      location               source  \
0 2022-08-05 13:54:28+00:00  she/they 25    Twitter for iPhone   
0 2022-08-05 11:24:59+00:00     Indonesia  Twitter for Android   
0 2022-08-05 05:47:06+00:00          None  Twitter for Android   
0 2022-08-04 17:17:57+00:00          None  Twitter for Android   
0 2022-08-04 15:27:59+00:00          None  Twitter for Android   

                                                 url  
0  https://twitter.com/gojotoji/status/1555552823...  
0  https://twitter.com/Kshatriya_666_/status/1555...  
0  https://twitter.com/Skullse1/status/1555430172...  
0  https://twitter.com/Skullse1/status/1555241643...  
0  https://twitter.com/Jackburger__/status/155521...

In [183]:
#Creating loop for word: war devil
wardevil_df = pd.DataFrame()
query4 = '(wardevil OR "war devil" OR #wardevil) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = 1555714802068201473

for i in range(6):
    tweets4 = client.search_recent_tweets(query=query4, max_results = 100, until_id = last_tw_id,
                                    user_fields = ['username', 'location'],
                                    tweet_fields=['created_at', 'source'],
                                    expansions = ['author_id'])
    temp_df = df_response(tweets4).sort_values(['created_at'], ascending=False)
    last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
        
    wardevil_df = pd.concat([temp_df, wardevil_df])

In [185]:
wardevil_df.shape

(526, 7)

In [187]:
wardevil_df.head()

tweet_id        username  \
0  1553422338590384131      DrawDemian   
0  1553411403180118017    MarkMononoke   
0  1553410607566684162    Hyper_Planar   
0  1553410098411732992         dnbooru   
0  1553396783471923201  Ebonylamblambb   

                                                text  \
0      Wardevil #chainsawman https://t.co/CGhTGM1tuH   
0  Cannot wait to see what the War devil is truly...   
0  FRAMED WAR DEVIL PRINT - RAFFLE ends today at ...   
0  World\n#chainsaw_man\nCharacters\n#mitaka_asa\...   
0  Asa Mitaka (csm)\n-\nre upload cuz i made some...   

                 created_at         location               source  \
0 2022-07-30 16:48:41+00:00             None  Twitter for Android   
0 2022-07-30 16:05:14+00:00       Austin, TX      Twitter Web App   
0 2022-07-30 16:02:04+00:00  San Antonio, TX   Twitter for iPhone   
0 2022-07-30 16:00:03+00:00            Tokyo         Danbooru Bot   
0 2022-07-30 15:07:08+00:00             None   Twitter for iPhone   

                                                 url  
0  https://twitter.com/DrawDemian/status/15534223...  
0  https://twitter.com/MarkMononoke/status/155341...  
0  https://twitter.com/Hyper_Planar/status/155341...  
0  https://twitter.com/dnbooru/status/15534100984...  
0  https://twitter.com/Ebonylamblambb/status/1553...

In [188]:
# merged and final dataframe
frames = [mitaka_df, asamitaka_df, mitakaasa_df, wardevil_df]
tw_mitaka_df = pd.concat(frames)
tw_mitaka_df.head()

#tw_mitaka_df = tw_mitaka_df[~tw_mitaka_df.index.duplicated(keep='first')] #Removing dup indices later

tweet_id       username  \
0  1553777739773186048      Vulpeqxla   
0  1553776451186458625      aokisotaa   
0  1553775430276194304  retromanjapan   
0  1553773879650394115      nymphahri   
0  1553756835047690240       moonrcin   

                                                text  \
0  i need to draw mitaka again soon or else i wil...   
0  finally decided to check out the denji edit ta...   
0  The Showa Retro In https://t.co/H3BK93ITy6 #wo...   
0  Asa Mitaka ⛓️\nShould I make the actual cospla...   
0  mitaka best girl \n\n#チェンソーマン #chainsawman htt...   

                 created_at      location               source  \
0 2022-07-31 16:20:55+00:00   they/them ♥   Twitter for iPhone   
0 2022-07-31 16:15:48+00:00  19, they/she  Twitter for Android   
0 2022-07-31 16:11:44+00:00  Tokyo, Japan    Revive Social App   
0 2022-07-31 16:05:35+00:00        Italia  Twitter for Android   
0 2022-07-31 14:57:51+00:00            🇵🇭   Twitter for iPhone   

                                                 url  
0  https://twitter.com/Vulpeqxla/status/155377773...  
0  https://twitter.com/aokisotaa/status/155377645...  
0  https://twitter.com/retromanjapan/status/15537...  
0  https://twitter.com/nymphahri/status/155377387...  
0  https://twitter.com/moonrcin/status/1553756835...

In [189]:
tw_mitaka_df.nunique()

tweet_id      846
username      694
text          830
created_at    845
location      409
source         18
url           846
dtype: int64

In [190]:
tw_mitaka_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1059 entries, 0 to 0
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype              
---  ------      --------------  -----              
 0   tweet_id    1059 non-null   int64              
 1   username    1059 non-null   object             
 2   text        1059 non-null   object             
 3   created_at  1059 non-null   datetime64[ns, UTC]
 4   location    698 non-null    object             
 5   source      1059 non-null   object             
 6   url         1059 non-null   object             
dtypes: datetime64[ns, UTC](1), int64(1), object(5)
memory usage: 66.2+ KB


In [191]:
tw_mitaka_df.to_csv('tw_mitaka.csv', index=False)

Since I can only query from the past 7 days, need to do more queries in future to get more datapoints.\
Would like to get at least 100k datapoints. 

8/5 Update: 846 unique tweets from 7/29 - 8/5\
8/12 Update: ### unique tweets from 8/5 - 8/12